File: scripts/navigation/point_navigation.py
#!/usr/bin/env python3
import rospy
import actionlib
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal

i = 0

my_2d_list = [
    [-2.16, 0.764],
    [-0.68, 1.6],
    [-0.55, -1.55],
    [-2,0]
]

def active_cb(extra):
    rospy.loginfo("Goal pose being processed")
    

def feedback_cb(feedback):
    rospy.loginfo("Current location: "+str(feedback))

def done_cb(status, result):
    if status == 3:
        rospy.loginfo("Goal reached")
    if status == 2 or status == 8:
        rospy.loginfo("Goal cancelled")
    if status == 4:
        rospy.loginfo("Goal aborted")
   

if __name__ == "__main__":
    rospy.init_node("point_nav")
    rospy.loginfo("Starting point_.")

    for i in range(len(my_2d_list)):
        navclient = actionlib.SimpleActionClient('move_base',MoveBaseAction)
        navclient.wait_for_server()

        # Example of navigation goal
        goal = MoveBaseGoal()
        goal.target_pose.header.frame_id = "map"
        goal.target_pose.header.stamp = rospy.Time.now()

        goal.target_pose.pose.position.x = my_2d_list[i][0]
        goal.target_pose.pose.position.y = my_2d_list[i][1]
        goal.target_pose.pose.position.z = 0.0
        goal.target_pose.pose.orientation.x = 0.0
        goal.target_pose.pose.orientation.y = 0.0
        goal.target_pose.pose.orientation.z = 0.662
        goal.target_pose.pose.orientation.w = 0.750

        i=i+1

        navclient.send_goal(goal, done_cb, active_cb, feedback_cb)
        finished = navclient.wait_for_result()

        if not finished:
            rospy.loginfo("Action server not available!")
        else:
            rospy.loginfo ( navclient.get_result())


    
    while not rospy.is_shutdown():
        pass

File: src/navigation/point_navigation.cpp
#include <ros/ros.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <actionlib/client/simple_action_client.h>


//global variables
float points[4][2] = {
    {-2.16, 0.764},
    {-0.68, 1.6},
    {-0.55, -1.55},
    {-2, 0}
};

typedef actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> MoveBaseClient;

void doneCallback(const actionlib::SimpleClientGoalState& state, const move_base_msgs::MoveBaseResultConstPtr& result) {
    ROS_INFO("Finished in state [%s]", state.toString().c_str());
    // You can perform additional actions or checks based on the result here

}

void activeCallback() {
    ROS_INFO("Goal just went active");
}

void feedbackCallback(const move_base_msgs::MoveBaseFeedbackConstPtr& feedback) {
    ROS_INFO("Got feedback of the current position - x: %f, y: %f", feedback->base_position.pose.position.x, feedback->base_position.pose.position.y);
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "point_navigation_cpp");
    ros::NodeHandle nh;
    

    for(int i = 0; i< sizeof(points); ++i)
    {
        MoveBaseClient ac("move_base", true);

        while (!ac.waitForServer(ros::Duration(5.0))) {
            ROS_INFO("Waiting for the move_base action server to come up");
        }

        move_base_msgs::MoveBaseGoal goal;
        goal.target_pose.header.frame_id = "map";
        goal.target_pose.header.stamp = ros::Time::now();

        goal.target_pose.pose.position.x = points[i][0];
        goal.target_pose.pose.position.y = points[i][1];
        goal.target_pose.pose.orientation.w = 1.0;

        // Set up callbacks
        ac.waitForServer();
        // ac.SimpleDoneCallback(boost::bind(&doneCallback, _1, _2));
        // ac.SimpleActiveCallback(boost::bind(&activeCallback));
        // ac.SimpleFeedbackCallback(boost::bind(&feedbackCallback, _1));

        ac.sendGoal(goal, doneCallback, activeCallback, feedbackCallback);

        ROS_INFO("Sending goal");
        // ac.sendGoal(goal);

        ac.waitForResult();

        if (ac.getState() == actionlib::SimpleClientGoalState::SUCCEEDED)
            ROS_INFO("Hooray, the base reached the goal!");
        else
            ROS_INFO("The base failed to reach the goal for some reason");

    }

    return 0;
}
